### PR TITLE
KVL-314 Add test for concurrent double spends

### PR DESCRIPTION
**Pull request by rautenrieth-da**
This PR adds a conformance test tool test that:
- Creates one contract and then concurrently archives it many times (10 tries across two participants)
- Repeats the above process 10 times

CHANGELOG_BEGIN
CHANGELOG_END

To check whether the test actually triggers the kvutils conflict detection, I have verified that the ledger log contains many lines like the following one:

```
11:53:59.732 ERROR c.d.l.v.b.ConflictDetection - Recovered a conflicted transaction, details='Conflict on contract 00d525918818d8782060b7c5ef2236f62c1370558ec5a198a6936980d8d366dd39' (context: {correlationId=SemanticConcurrentDoubleSpend-alpha-308a4414504b-command-39})
```

### CODE REVIEW

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala [hunk 1]
```diff
@@ -58,6 +59,44 @@ final class SemanticTests extends LedgerTestSuite {
...
+                    .exercise(owner2, shared.exerciseSharedContract_Consume2)
+                    .transform(Success(_))
+            })
+          } yield {
+            assertLength(s"Contract $c successful archives", 1, results.collect({
+              case Success(_) => true
+            }))
+            assertLength(s"Contract $c failed archives", archives - 1, results.collect({
+              case Failure(_) => true
+            }))
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala:94]: I think the builtin methods will help you here.

```suggestion
            assertLength(s"Contract $c successful archives", 1, results.filter(_.isSuccess))
            assertLength(s"Contract $c failed archives", archives - 1, results.filter(_.isFailure)))
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala [hunk 2]
```diff
@@ -58,6 +59,44 @@ final class SemanticTests extends LedgerTestSuite {
...
+              i % 2 match {
+                case 0 =>
+                  alpha
+                    .exercise(owner1, shared.exerciseSharedContract_Consume1)
+                    .transform(Success(_))
+                case 1 =>
+                  beta
+                    .exercise(owner2, shared.exerciseSharedContract_Consume2)
+                    .transform(Success(_))
+            })
```
Comment (SamirTalwar-DA) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala:87]: You could potentially make use of `Seq.fill` here to make this a bit more declarative, though I'm not sure whether it's bettor not. Something like:

```suggestion
            exercises = Seq.fill(archives / 2)(Seq(
              (alpha, owner1, shared.exerciseSharedContract_Consume1),
              (beta, owner2, shared.exerciseSharedContract_Consume2),
            )).flatten
            results <- Future.traverse(exercises) {
              case (participant, owner, consume) =>
                participant.exercise(owner, consume).transform(Success(_))
            }
```

_I_ prefer it, but you don't have to agree.

**Review by SamirTalwar-DA — APPROVED** [review]
The test looks good to me. I've got some syntactic suggestions, but nothing important.

**Comment by SamirTalwar-DA** [comment]
Side note: that conflict detection error probably shouldn't be an error; it's expected behavior that's out of the operator's control. I'd prefer it was an INFO.

**Review by fabiotudone-da — APPROVED** [review]
LGTM

**Comment by miklos-da** [comment]
> Side note: that conflict detection error probably shouldn't be an error; it's expected behavior that's out of the operator's control. I'd prefer it was an INFO.

Agreed -- will change it to INFO.

**Pull request merged**